### DISCUSSION TITLE
EnvKey and Terraform

### DISCUSSION THREAD

**Discussion by robertvilvert** [envkey/envkey#75] [Q&A]
Is it possible to integrate EnvKey with Terraform for storing sensitive data? Is there already a module for this?

**Comment by danenania** [comment]
@robertvilvert and I ended up taking this discussion to email, but I wanted to also post a response here in case it's useful to others 👇

Yes, you can use EnvKey with Terraform. While we don't have a specific plugin for Terraform, you just need to be able to set an environment variable with an ENVKEY (access token) on your server to integrate. Since terraform configs are usually in source control, we wouldn't recommend setting an ENVKEY directly here in any case because these are sensitive. The best way to store the ENVKEY and expose it as an environment variable would depend on your hosting platform.

On GCP or AWS, for example, the typical path would be to generate an ENVKEY for each of your environments, store it as a secret in AWS/GCP secrets manager, and then inject the secret as an ENVKEY environment variable to your container or process.

In terraform, it would look something like this (this is a GCP example but AWS would be similar):

```
resource "google_cloud_run_service" "default" {
  name     = "my-service"
  location = "us-central1"

  template {
    spec {
      containers {
        image = "[gcr.io/project-id/my-image](http://gcr.io/project-id/my-image)"
        env {
          name = "ENVKEY"
          value_from {
            secret_key_ref {
              name = google_secret_manager_secret.my_app_production_envkey.secret_id
              key  = "latest"
            }
          }
        }
      }
    }
  }

  traffic {
    percent         = 100
    latest_revision = true
  }
}
```

Lastly, I should note that we are focusing on integrations currently so if people think an official terraform module would be an improvement on the above approach, please chime in!
